### DISCUSSION TITLE
Debugger support in notebook cells

### DISCUSSION THREAD

**Discussion by profix898** [dotnet/interactive#1596] [General]
Is there any way to use a debugger (Omnisharp) inside a C# (or Powershell) cell to step through the code, set breakpoints, and inspect the variables?

If no, are there any plans to support that scenario?

I really like the dotNet Interactive experience inside of VSCode. The fact that you can mix code and documentation and have formatted output (text, table, graphics or html) embedded in the notebook. But the developer experience is pretty bad.
1. As reported in #1391, code/symbol completion is extremely slow (typically several seconds, took a while to figure out that it actually has code completion at all)
2. No variables displayed as reported in #1581
3. No debugger support inside notebooks cells (this discussion)

Taking all these together (and with the last point weighting in), it is very difficult to get your code right. Especially for C# beginners or intermittent users.

**Comment by jonsequitur** [comment]
A step-through debugger is something we're considering, though it's not in the classic notebook feature set so it's not something we've prioritized so far. Jupyter released support for it last year and we'll likely support it at some point. One thing about a debugger is that it's only helpful for beginners if they already know how to use a debugger, so we're also looking at more intuitive ways to make the state of a running program visible. 

The latest .NET Interactive release has made some gains on #1391. Please give it a try and let us know if you see any improvement.

**Reply by 2Dews** [reply]
Has there been any further consideration for a step-through debugging experience for notebook cells? I use notebooks for prototyping and it would be very helpful. I use it regularly in VS Code notebooks with Python and its very helpful when exploring data.

**Comment by profix898** [comment]
Thanks for your comment on the plans/status of the debugger support. Yes, people should know how to use a debugger. In my world (university), students usually know that, but they start using dotNet Interactive to play around and learn .NET and the C# language. It's a low-barrier entry into the .NET world without installing GB-sized VisualStudio. And for them it is pretty essential to have all three components I mentioned above.
The second scenario I am facing are people with (basic) C# knownledge, who want to only consume a library in a few lines of code, e.g. to evaluate scientific data or generate a graph intermixed with documentation. They usually only want to install the full VS if they start contributing to the libraries. However, for now I think I should advise them to create a regular C# source file in VS Code (with Omnisharp). But I'm looking forward to using dotNet Interactive for that in the (near) future.

On #1391:
Not sure that helped. If you load a few nuget packages, it still takes serveral seconds (typically 6-8 secs in my tests) until the auto completion shows up. My impression is that it might be a little faster when you complete on something you have explored before (still takes >2 secs though). Attached is a small screencast of my code completion experience. It uses the [ILNInteractive](https://github.com/profix898/ILNInteractive) extension for dotNet Interactive to bring scientific computing/visualization into VS Code.

![complete_delay](https://user-images.githubusercontent.com/2789426/133331778-5f3afb82-e7e7-4acd-b425-17650171ffb3.gif)
